W-20897660: Forward compatibility (#298)
mulesoft/docs-mule-sdk: modules/ROOT/pages/forward-compatibility-patterns.adoc

= Forward Compatibility Patterns
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: mule, sdk, forward compatibility, RuntimeVersion, MuleVersion, conditional behavior

When building forward-compatible modules, you often need to use newer SDK features while providing fallback behavior for older runtimes. The following patterns show how to implement conditional behavior based on the Mule runtime version.

== Check the Mule Runtime Version

Use the `@RuntimeVersion` annotation to inject the current Mule Runtime version into your component. You can apply this annotation to fields of type `org.mule.runtime.api.meta.MuleVersion`.

> NOTE: The `mule-sdk-api` artifact has a dependency on `mule-api` for basic APIs that are not intended to evolve over time. `MuleVersion` is one of these stable APIs, so there is no `org.mule.sdk.api` counterpart for it. You can safely use `org.mule.runtime.api.meta.MuleVersion` in forward compatible modules.

Reviewer: ```suggestion
NOTE: The `mule-sdk-api` artifact has a dependency on `mule-api` for basic APIs that aren't intended to evolve over time. `MuleVersion` is one of these stable APIs, so there's no `org.mule.sdk.api` counterpart for it. You can safely use `org.mule.runtime.api.meta.MuleVersion` in forward compatible modules.
```

[source,java,linenums]
----
import org.mule.runtime.api.meta.MuleVersion;
import org.mule.sdk.api.annotation.RuntimeVersion;

public class MyOperations {

  @RuntimeVersion
  private MuleVersion runtimeVersion;

  public void myOperation() {
    if (runtimeVersion.atLeast("4.12.0")) {
      // Use newer feature
    } else {
      // Use fallback behavior
    }
  }
}
----

The `MuleVersion` class provides methods to compare versions:

[%header,cols="1,2"]
|===
| Method | Description

| `atLeast(String version)`
| Returns `true` if the Mule runtime version is equal to or greater than the specified version.

| `atLeastBase(String version)`
| Returns `true` if the base version (ignoring patch level) is equal to or greater than the specified version.
|===

The runtime version is constant across your module, so @RuntimeVersion applies only at the field level. Centralize your version-dependent logic in a shared field rather than checking in every operation.

== Call New Methods in Existing APIs

> SDK APIs may add new methods in later versions. If you call a method that doesn't exist on an earlier Mule runtime version, you get a `NoSuchMethodError`.

Reviewer: ```suggestion
SDK APIs can add new methods in later versions. If you call a method that doesn't exist on an earlier Mule runtime version, you get a `NoSuchMethodError`.
```
